### PR TITLE
Commit annotations

### PR DESCRIPTION
**Pull request by EduardSergeev**
Generic mechanism for annotating `ivory` commits with a set of key=value pairs.

First intended use is to store `napoleon`s commits corresponding to exported dictionaries.

### CODE REVIEW

File: ivory-cli/src/main/scala/com/ambiata/ivory/cli/IvoryCmd.scala [hunk 1]
```diff
@@ -79,14 +79,15 @@ object IvoryCmd {
...
+      case MetadataVersion.V0 | MetadataVersion.V1 | MetadataVersion.V2 | MetadataVersion.V3 |
+           MetadataVersion.V4 | MetadataVersion.V5 | MetadataVersion.V6 | MetadataVersion.V7 =>
         RIO.failIO[Unit](s"""The version of the ivory repository you are trying to access has
                                 |meta-data in a form which is too old to be read, you need to run
                                 |run `ivory update` in order for this version of ivory to proceed.
                                 |
                                 |WARNING: If you run `ivory update` older ivory installs will no
                                 |longer be able to access the repository.""".stripMargin)
-      case MetadataVersion.V7 =>
+      case MetadataVersion.V8 =>
```
Comment (charleso) [inline ivory-cli/src/main/scala/com/ambiata/ivory/cli/IvoryCmd.scala]: You'll need to change https://github.com/ambiata/ivory/blob/master/ivory-cli/src/test/bin/tests/data-regression/expected/metadata/config#L2 this as well to get the cli tests to pass

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/AnnotationTextStorage.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    val key = Repository.annotation(id)
+    for {
+      exists <- repository.store.exists(key)
+      annotation <-
+      if (exists)
+        for {
+          lines <- repository.store.utf8.read(key)
+          result <- RIO.fromDisjunctionString[Annotation](Parse.decodeEither[Annotation](lines))
+        } yield result
+      else Annotation.empty.point[RIO]
```
Comment (charleso) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/AnnotationTextStorage.scala]: I'm not sure this is right (but apologies if I've done something like this elsewhere). If you ask for an annotation that doesn't exist it should return `None`. One suggestion would be to make `annotationSetId` `Option[AnnotationId]` for legacy repositories.

Comment (EduardSergeev) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/AnnotationTextStorage.scala]: Yes, I've just realised that it was wrong (when I tried to plug it into cli). Will make it identical to other *Storages' behaviour.

**Comment by charleso** [comment]
Just bikeshedding with names, but how does "notes" sound rather than "annotations"?

File: ivory-cli/src/test/bin/tests/commits/test [hunk 3]
```diff
@@ -18,3 +18,6 @@ commit_test 00000002
 
 $IVORY import-dictionary -p "${INPUT}/dictionary.psv"
 commit_test 00000003
+
+$IVORY update-annotation -a napoleon-commit=00001234
```
Comment (charleso) [inline ivory-cli/src/test/bin/tests/commits/test]: I might be tempted to have an `annotation` command with `update` as a sub-sub command. Thoughts?

Either way be good to have a `get-annotation` command for testing the value of `napoleon-commit` here

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/annotation/AnnotationManager.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package com.ambiata.ivory.operation.annotation
+
+import com.ambiata.ivory.core.{CommitId, Annotation, Repository}
+import com.ambiata.ivory.storage.control._
+import com.ambiata.ivory.storage.metadata.{Metadata, AnnotationTextStorage}
+import com.ambiata.mundane.control.RIO
+
+
+object AnnotationManager {
```
Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/annotation/AnnotationManager.scala]: We would normally just call this `Annotation` :)

Comment (EduardSergeev) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/annotation/AnnotationManager.scala]: It clashes with [this Annotation](https://github.com/ambiata/ivory/blob/topic/commit-annotation/ivory-core/src/main/scala/com/ambiata/ivory/core/Annotation.scala) then (inspiration for name came to me from [DictionaryImporter](https://github.com/ambiata/ivory/blob/master/ivory-operation/src/main/scala/com/ambiata/ivory/operation/ingestion/DictionaryImporter.scala)). Any other suggestion for name or how avoid the clash?

**Comment by EduardSergeev** [comment]
> Just bikeshedding with names, but how does "notes" sound rather than "annotations"?

Sounds shorter :) I'd even prefer "tags" but I think this name didn't play well with some other "tags".

**Comment by EduardSergeev** [comment]
@charleso @raronson Minimal but complete (for napoleon needs so far) implementation.

PS I kind of still like `annotation` vs `note`, but willing to change if other people prefer it (or something else).

**Comment by markhibberd** [comment]
@EduardSergeev I am :+1: for annotation.

File: ivory-cli/src/main/scala/com/ambiata/ivory/cli/annotation/get.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    |""".stripMargin)
+
+  , (     flag[CommitId](both('c', "commit"), empty)
+      |@| IvoryCmd.repository
+
+  )((commitId, repo) => IvoryRunner(configuration => repo(configuration).flatMap(repository =>
+
+      IvoryT.fromRIO(for {
+          annotation <- AnnotationManager.getAnnotation(repository, commitId)
+          result = annotation.map(Annotation.toJson(_)).getOrElse("")
```
Comment (charleso) [inline ivory-cli/src/main/scala/com/ambiata/ivory/cli/annotation/get.scala]: I wonder if you should just return an empty List from this command?

**Comment by charleso** [comment]
:+1:

File: ivory-cli/src/main/scala/com/ambiata/ivory/cli/annotation/get.scala (hunk 5 above)
Comment (EduardSergeev) [inline ivory-cli/src/main/scala/com/ambiata/ivory/cli/annotation/get.scala]: Yes, actually I was meant to ask people's opinion on how who we handle missing annotations. `"{}"`, `""` or some other indication? Empty JSON seems ok to me: simpler handling on client side too (do we want to distinguish the cases of empty annotation vs absent one?).

Comment (charleso) [inline ivory-cli/src/main/scala/com/ambiata/ivory/cli/annotation/get.scala]: Yeah I don't know. Returning `{}` seems fine for now.

**Comment by EduardSergeev** [comment]
Bump @markhibberd `<|>` @raronson `<|>` @nhibberd  OK to merge?

**Comment by nhibberd** [comment]
Will defer to @markhibberd @raronson

File: ivory-cli/src/main/scala/com/ambiata/ivory/cli/IvoryCmd.scala [hunk 6]
```diff
@@ -10,7 +10,7 @@ import java.util.UUID
 
 import org.apache.hadoop.fs.Path
 
-import pirate._, Pirate._
+  import pirate._, Pirate._
```
Comment (markhibberd) [inline ivory-cli/src/main/scala/com/ambiata/ivory/cli/IvoryCmd.scala]: Accidental space?

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/CommitStorage.scala [hunk 7]
```diff
@@ -9,10 +9,11 @@ import scalaz._, Scalaz._
 import com.ambiata.mundane.control._
 import com.ambiata.ivory.core._
 
-// 3 lines.
+// 4 lines.
 // The first line is dictionaryId: DictionaryId
 // The second line is featureStoreId: FeatureStoreId
 // The third line is an optional repositoryConfigId: RepositoryConfigId
+// The fourth line is an optional set of annotations : Map[String, Annotation]
```
Comment (markhibberd) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/CommitStorage.scala]: Assuming here, repository-config is mandatory if annotation line is set?

File: ivory-cli/src/test/bin/tests/annotation/test [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+$IVORY import-dictionary -p "${INPUT}/dictionary.psv"
+commit_test 00000003
+
+$IVORY annotation get -c 00000003 > "${TARGET}/empty.json"
+annotation_test empty.json
+
+$IVORY annotation update -a napoleon-commit=00001234
+commit_test 00000004
+
+$IVORY annotation get -c 00000004 > "${TARGET}/annotation.json"
```
Comment (markhibberd) [inline ivory-cli/src/test/bin/tests/annotation/test]: Would be good to add an extra check to this, for the case where the next ingest happens, that maintains the annotation.

**Comment by markhibberd** [comment]
:+1: After extra check added to cli (if that comment makes sense)

**Comment by markhibberd** [comment]
@EduardSergeev Actually, can you grab @raronson and myself this afternoon to discuss this one. Want to make sure we all understand the flow.

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/CommitStorage.scala (hunk 7 above)
Comment (EduardSergeev) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/metadata/CommitStorage.scala]: Must be, being the fourth line (after the third one). Do you mean it should be expressed here more explicitly? Or is it possible for new repositories to not have "repository config" line?

**Pull request merged**